pkg/util/lifted/scheduler/cache/node_tree_test.go

```diff
@@ -0,0 +1,409 @@
+		},
+		{
+			name: "remove last node in zone",
+			existingNodes: []*corev1.Node{
+				makeNode("node1", "us-east", "us-east-1a"),
+			},
+			nodeToRemove:  makeNode("node1", "us-east", "us-east-1a"),
+			expectedZones: []string{},
+			expectedCount: 0,
+			expectedError: false,
+		},
+		{
+			name: "remove non-existing node",
+			existingNodes: []*corev1.Node{
+				makeNode("node1", "us-east", "us-east-1a"),
+			},
+			nodeToRemove: makeNode("node2", "us-east", "us-east-1a"),
+			expectedZones: []string{
+				getExpectedZoneKey("us-east", "us-east-1a"),
+			},
+			expectedCount: 1,
+			expectedError: true,
+		},
+		{
+			name: "remove unlabeled node",
+			existingNodes: []*corev1.Node{
+				{ObjectMeta: metav1.ObjectMeta{Name: "node1"}},
+			},
+			nodeToRemove:  &corev1.Node{ObjectMeta: metav1.ObjectMeta{Name: "node1"}},
+			expectedZones: []string{},
+			expectedCount: 0,
+			expectedError: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			nt := newNodeTree(tt.existingNodes)
+			err := nt.removeNode(tt.nodeToRemove)
+
+			if tt.expectedError {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
+
+			assert.ElementsMatch(t, tt.expectedZones, nt.zones)
+			assert.Equal(t, tt.expectedCount, nt.numNodes)
+		})
+	}
+}
+
+func TestUpdateNode(t *testing.T) {
+	tests := []struct {
+		name          string
+		existingNodes []*corev1.Node
+		oldNode       *corev1.Node
+		newNode       *corev1.Node
+		expectedZones []string
+		expectedCount int
+	}{
+		{
+			name: "update to same zone",
+			existingNodes: []*corev1.Node{
+				makeNode("node1", "us-east", "us-east-1a"),
+			},
+			oldNode:       makeNode("node1", "us-east", "us-east-1a"),
+			newNode:       makeNode("node1", "us-east", "us-east-1a"),
+			expectedZones: []string{getExpectedZoneKey("us-east", "us-east-1a")},
+			expectedCount: 1,
+		},
+		{
+			name: "update to different zone",
+			existingNodes: []*corev1.Node{
+				makeNode("node1", "us-east", "us-east-1a"),
+			},
+			oldNode:       makeNode("node1", "us-east", "us-east-1a"),
+			newNode:       makeNode("node1", "us-east", "us-east-1b"),
+			expectedZones: []string{getExpectedZoneKey("us-east", "us-east-1b")},
+			expectedCount: 1,
+		},
+		{
+			name: "update from unlabeled to labeled",
+			existingNodes: []*corev1.Node{
+				{ObjectMeta: metav1.ObjectMeta{Name: "node1"}},
+			},
+			oldNode:       &corev1.Node{ObjectMeta: metav1.ObjectMeta{Name: "node1"}},
+			newNode:       makeNode("node1", "us-east", "us-east-1a"),
+			expectedZones: []string{getExpectedZoneKey("us-east", "us-east-1a")},
+			expectedCount: 1,
+		},
+		{
+			name: "update from labeled to unlabeled",
+			existingNodes: []*corev1.Node{
+				makeNode("node1", "us-east", "us-east-1a"),
+			},
+			oldNode:       makeNode("node1", "us-east", "us-east-1a"),
+			newNode:       &corev1.Node{ObjectMeta: metav1.ObjectMeta{Name: "node1"}},
+			expectedZones: []string{""},
+			expectedCount: 1,
+		},
+	}
```

Comment:
We can add a test case： updates a node in the NodeTree when the old node does not exist.